test_suite/tests/ui/pass_compact_generic_with_unbounded_trait_impl.rs

```diff
@@ -0,0 +1,33 @@
+use scale_info::TypeInfo;
+use scale::{Codec, Encode, Decode};
+
+#[derive(TypeInfo, Encode, Decode)]
+struct Foo<T> {
+    // When the below attribute is uncommented it will trigger the error
+    //   `overflow evaluating the requirement `_::_parity_scale_codec::Compact<_>: Decode`
+    // It will expand to a call of `.compact_of::<T>()` which triggers the overflow.
+    //
+    #[codec(compact)]
+    b: T,
+}
+
+trait Trait {
+    type Type: TypeInfo + 'static;
+}
+
+#[derive(TypeInfo, Encode, Decode)]
+struct Bar<T>(std::marker::PhantomData<T>);
+
+impl<T> Trait for Bar<T>
+where
+    T: TypeInfo + 'static,
+    // Foo<T>: TypeInfo + 'static, // adding this bound will fix the overflow error:
```

Comment:
Without this bound we should get a compilation error, instead we get infinite recursion.